cluster-autoscaler/cloudprovider/gce/autoscaling_gce_client.go

```diff
@@ -166,22 +168,77 @@ func (client *autoscalingGceClientV1) DeleteInstances(migRef GceRef, instances [
 	return client.waitForOp(op, migRef.Project, migRef.Zone)
 }
 
-func (client *autoscalingGceClientV1) FetchMigInstances(migRef GceRef) ([]GceRef, error) {
-	registerRequest("instance_group_managers", "list_managed_instances")
-	instances, err := client.gceService.InstanceGroupManagers.ListManagedInstances(migRef.Project, migRef.Zone, migRef.Name).Do()
+func (client *autoscalingGceClientV1) FetchMigInstances(migRef GceRef) ([]cloudprovider.Instance, error) {
+	gceInstances, err := client.gceService.InstanceGroupManagers.ListManagedInstances(migRef.Project, migRef.Zone, migRef.Name).Do()
 	if err != nil {
 		klog.V(4).Infof("Failed MIG info request for %s %s %s: %v", migRef.Project, migRef.Zone, migRef.Name, err)
 		return nil, err
 	}
-	refs := []GceRef{}
-	for _, i := range instances.ManagedInstances {
-		ref, err := ParseInstanceUrlRef(i.Instance)
+	infos := []cloudprovider.Instance{}
+	for _, gceInstance := range gceInstances.ManagedInstances {
+		ref, err := ParseInstanceUrlRef(gceInstance.Instance)
 		if err != nil {
 			return nil, err
 		}
-		refs = append(refs, ref)
+
+		instance := cloudprovider.Instance{
+			Id:     ref.ToProviderId(),
+			Status: &cloudprovider.InstanceStatus{},
+		}
+
+		switch gceInstance.CurrentAction {
+		case "CREATING", "RECREATING", "CREATING_WITHOUT_RETRIES":
+			instance.Status.State = cloudprovider.InstanceCreating
+		case "ABANDONING", "DELETING":
+			instance.Status.State = cloudprovider.InstanceDeleting
+		default:
+			instance.Status.State = cloudprovider.InstanceRunning
+		}
+
+		if instance.Status.State == cloudprovider.InstanceCreating {
+			var errorInfo cloudprovider.InstanceErrorInfo
+			errorMessages := []string{}
+			errorFound := false
+			for _, instanceError := range getLastAttemptErrors(gceInstance) {
+				if isStockoutErrorCode(instanceError.Code) {
+					errorInfo.ErrorClass = cloudprovider.OutOfResourcesErrorClass
+					errorInfo.ErrorCode = "STOCKOUT"
+				} else if isQuotaExceededErrorCoce(instanceError.Code) {
+					errorInfo.ErrorClass = cloudprovider.OutOfResourcesErrorClass
+					errorInfo.ErrorCode = "QUOTA_EXCEEDED"
```

Comment:
Can we make STOCKOUT and QUOTA_EXCEEDED consts? It's likely we'll need to refer to them at some point.